### PR TITLE
add_n & add_continous_stat for Likert tables

### LINKED ISSUE #55: `tbl_likert()` features
- [ ] Improve `add_n()` to allow for more statistics (e.g. missing N or percent).
- [ ] Add functionality to include the continuous summary statistics.

### PR DESCRIPTION
**Pull request by larmarange**
fix #55

- [ ] add examples in documentation
- [ ] add tests

### CODE REVIEW

**Comment by larmarange** [comment]
Hi @ddsjoberg 

I hope you are well. I have rewritten `add_n.tbl_likert()` to use all functionalities available in `add_n.tbl_summary`.

I have also added a `add_continuous_stat()` method (name could be changed) to allow the addition of continuous statistics from a Likert table. You may apply `add_continuous_stat()` several times.

``` r
library(tidyverse)
library(bstfun)

data(efc, package = "sjPlot")

cop <- sjmisc::find_var(efc, pattern = "cop", out = "df") %>%
  sjmisc::numeric_to_factor() %>%
  mutate(across(everything(), ~ fct_relabel(.x, str_to_lower))) %>%
  as_tibble() %>%
  mutate(c82cop1 = labelled::recode_if(c82cop1, c82cop1 == "never", NA))
names(cop) <- paste0("q", 1:ncol(cop))

cop %>% 
  tbl_likert(statistic = ~ "{p}%", sort = "desc") %>%
  add_n(statistic = "{n}/{N}", footnote = TRUE) %>%
  add_continuous_stat() %>%
  add_continuous_stat(statistic = "{sd}") %>%
  as_kable()
```

| **Characteristic**                                                         | **N**   | **always** | **never** | **often** | **sometimes** | **Mean** | **SD** |
|:---------------------------------------------------------------------------|:--------|:-----------|:----------|:----------|:--------------|:---------|:-------|
| do you find caregiving too demanding?                                      | 902/908 | 4.3%       | 21%       | 14%       | 61%           | 3.31     | 0.94   |
| do you feel trapped in your role as caregiver?                             | 900/908 | 8.6%       | 37%       | 13%       | 42%           | 2.87     | 1.06   |
| does caregiving have negative effect on your physical health?              | 898/908 | 6.5%       | 46%       | 9.5%      | 39%           | 2.80     | 1.03   |
| do you feel supported by friends/neighbours?                               | 901/908 | 12%        | 35%       | 27%       | 26%           | 2.67     | 1.00   |
| do you feel you cope well as caregiver?                                    | 898/908 | 23%        | 0%        | 66%       | 11%           | 2.64     | 0.96   |
| does caregiving cause difficulties in your relationship with your friends? | 902/908 | 5.8%       | 57%       | 9.1%      | 28%           | 2.59     | 0.96   |
| does caregiving cause difficulties in your relationship with your family?  | 902/908 | 1.7%       | 69%       | 5.5%      | 23%           | 2.51     | 0.87   |
| do you feel caregiving worthwhile?                                         | 888/908 | 34%        | 8.6%      | 34%       | 24%           | 2.47     | 1.18   |
| does caregiving cause financial difficulties?                              | 900/908 | 1.9%       | 79%       | 4.3%      | 15%           | 2.32     | 0.74   |

<sup>Created on 2021-10-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by larmarange** [comment]
For a Likert table, I have the feeling that `"{p}%"` would be a better and more common default value for `statistic`.

**Comment by larmarange** [comment]
I have updated the example (levels order was incorrect). In this example, the last variable has voluntarily a missing level.

Should we think about an example dataset to provide for `tbl_likert`?

``` r
library(tidyverse)
library(bstfun)

data(efc, package = "sjPlot")

cop <- sjmisc::find_var(efc, pattern = "cop", out = "df") %>%
  sjmisc::numeric_to_factor() %>%
  mutate(across(everything(), ~ fct_relabel(.x, str_to_lower) %>% fct_relevel("never", "sometimes", "often", "always"))) %>%
  as_tibble() %>%
  mutate(c82cop1 = labelled::recode_if(c82cop1, c82cop1 == "never", NA) %>% fct_drop()) %>%
  relocate(c82cop1, .after = last_col())
names(cop) <- paste0("q", 1:ncol(cop))

cop %>% 
  tbl_likert(statistic = ~ "{p}%", sort = "desc") %>%
  add_n(statistic = "{n}/{N}", footnote = TRUE) %>%
  add_continuous_stat() %>%
  add_continuous_stat(statistic = "{sd}") %>%
  as_kable()
```

| **Characteristic**                                                         | **N**   | **never** | **sometimes** | **often** | **always** | **Mean** | **SD** |
|:---------------------------------------------------------------------------|:--------|:----------|:--------------|:----------|:-----------|:---------|:-------|
| do you feel you cope well as caregiver?                                    | 898/908 | 0%        | 11%           | 66%       | 23%        | 2.13     | 0.57   |
| do you feel caregiving worthwhile?                                         | 888/908 | 8.6%      | 24%           | 34%       | 34%        | 2.93     | 0.96   |
| do you feel supported by friends/neighbours?                               | 901/908 | 35%       | 26%           | 27%       | 12%        | 2.16     | 1.04   |
| do you find caregiving too demanding?                                      | 902/908 | 21%       | 61%           | 14%       | 4.3%       | 2.02     | 0.72   |
| do you feel trapped in your role as caregiver?                             | 900/908 | 37%       | 42%           | 13%       | 8.6%       | 1.92     | 0.91   |
| does caregiving have negative effect on your physical health?              | 898/908 | 46%       | 39%           | 9.5%      | 6.5%       | 1.77     | 0.87   |
| does caregiving cause difficulties in your relationship with your friends? | 902/908 | 57%       | 28%           | 9.1%      | 5.8%       | 1.63     | 0.87   |
| does caregiving cause difficulties in your relationship with your family?  | 902/908 | 69%       | 23%           | 5.5%      | 1.7%       | 1.39     | 0.67   |
| does caregiving cause financial difficulties?                              | 900/908 | 79%       | 15%           | 4.3%      | 1.9%       | 1.29     | 0.64   |

<sup>Created on 2021-10-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: R/tbl_likert.R [hunk 1]
```diff
@@ -130,37 +140,198 @@ tbl_likert <- function(data,
...
 
-  # add n column ---------------------------------------------------------------
+  data <- x$inputs$data
+  include <- x$inputs$include
+
+  if ("n" %in% names(x$table_body))
+    stop("`add_n()` has already been applied", call. = FALSE)
+
+  # compute stat ---------------------------------------------------------------
+  data <-
```
Comment (ddsjoberg) [inline R/tbl_likert.R:164]: In `tbl_likert()` we used `fct_unify()` to unify all the levels of the variables, but here that is not being taken into account.

File: R/tbl_likert.R [hunk 2]
```diff
@@ -130,37 +140,198 @@ tbl_likert <- function(data,
...
+                             ...) {
   # check inputs ---------------------------------------------------------------
   if (!inherits(x, "tbl_likert"))
     stop("`x=` must be class 'tbl_likert'", call. = FALSE)
 
-  # add n column ---------------------------------------------------------------
+  data <- x$inputs$data
+  include <- x$inputs$include
+
+  if ("n" %in% names(x$table_body))
```
Comment (ddsjoberg) [inline R/tbl_likert.R:160]: To make this consistent with the other gtsummary functions, we should update `tbl_likert()` to return the named list of calls, and I think the added N column is called `"stat_n"`?

File: R/tbl_likert.R (hunk 1 above)
Comment (larmarange) [reply R/tbl_likert.R:164]: Would it be easier and more consistent to save the unified version of `data` as proposed here: https://github.com/ddsjoberg/bstfun/pull/62/commits/048045cd6afbe82e78552427b913169da9a66f57 ?

Comment (larmarange) [reply R/tbl_likert.R:166]: For these functions, it is not an issue, as we are just adding additional stat col but not modifying the labelling

**Comment by ddsjoberg** [comment]
hey hey @larmarange ! hope you're well! Am I meant to do something with this, or are you working on it?

**Comment by larmarange** [comment]
Dear @ddsjoberg   I'm truly sorry. I'm currently overloaded with my regular work and I forgot to finalise the PR. If you have time, do not hesitate to work on it / finalise it.

The code is working. In the discussion, there are some remaining questions:

- For a Likert table, I have the feeling that `"{p}%"` would be a better and more common default value for `statistic`.
- Should we think about an example dataset to provide for tbl_likert?
- We could/should provide more examples in documentation.
- Unit tests are missing

Best regards

PS: another open question to be addressed in a separate PR: should we add a `group` option in `tbl_likert()` to produce sub-tables according to a factor and to stack them? If yes, that argument will have to be taken into account also in `add_n()` and `add_continuous_stat()`

**Comment by ddsjoberg** [comment]
@larmarange no worries, and on apology needed 😸 

> PS: another open question to be addressed in a separate PR: should we add a group option in `tbl_likert()` to produce sub-tables according to a factor and to stack them? If yes, that argument will have to be taken into account also in `add_n()` and `add_continuous_stat()`

Would this functionality be covered in `tbl_strata()`?

**Comment by larmarange** [comment]
> Would this functionality be covered in `tbl_strata()`?

Yes it can be. Maybe a good idea to simply add an example with it.

**Comment by larmarange** [comment]
Let me know if you want to finalize the PR. Otherwise, I can try to find some time next week.

Regards

**Comment by ddsjoberg** [comment]
> Let me know if you want to finalize the PR. Otherwise, I can try to find some time next week.
> 
> Regards

I won't have time in the next week either. Go ahead when you've got the time. Happy (American) Thanksgiving! 🦃 🦃 🦃

**Comment by ddsjoberg** [comment]
hey hey @larmarange ! I hope you've been well :)

I may have some time to work on this in the first or second week of April. FYI If you want to work on it, let me know and I won't touch it.

**Comment by larmarange** [comment]
Hey. :-)

I'm completely overloaded until mid-April. So do not hesitate to move forward. And thanks for all you are doing.

**Review by ddsjoberg — APPROVED** [review]
👍🏼

**Comment by larmarange** [comment]
Thanks

**Pull request merged**